### ISSUE TITLE
Enhancement : option to execute actions sequentially

### ISSUE THREAD

**Issue by fbaligand** [lmangani/sentinl#99]
SENTINL is great to generate alerts and reports.
But with multiple webhook actions, you could do more ! 
We can, for example, automate some optimizations on yesterday logstash index (best_compression, force merge to one segment, ...)

And to do such a thing, it would be great to add a new option that let execute watcher actions sequentially (and not in parallel) !

**Comment by lmangani** [comment]
@fbaligand good point and good idea. How about the approach of setting up "exchange" indices where some actions write and other read their results (as in the ATLAS example) for this purpose? If not, could you outline how you envision the programming of their execution? One watcher launching another?

**Comment by fbaligand** [comment]
Hi @lmangani,

First, nice to see you find it's a good idea !

Then, my envision is not to execute several watchers sequentially, but rather to define only one watcher with several actions that are executed sequentially. For example, using javascript promises to chain actions.

I imagine two ways to configure it :

- first way : enable "sequential" option in actions, and all actions are executed sequentially, either using alphabetical action name order, either using definition order in json (not sure it's technically possible) :
``` json
{
    ...
    "actions": {
      "sequential": true,
      "my_webhook1": {
        "webhook": {
             ...
        }
      },
      "my_webhook2": {
        "webhook": {
             ...
        }
      }
    }
}
```

- second way : define "order" option in each action, and all actions are executed sequentially, using the configured order :
``` json
{
    ...
    "actions": {
      "my_webhook1": {
        "order": 1,
        "webhook": {
             ...
        }
      },
      "my_webhook2": {
        "order": 2,
        "webhook": {
             ...
        }
      }
    }
}
```

**Comment by chenryn** [comment]
I think it's much better to cover this use case by supporting script actions? Then users can implement their own sequential tasks in one script file.

**Comment by fbaligand** [comment]
I think these are 2 different interesting enhancements : 
- add a new "script_file" action
- add a new "sequential" action option

For the use case I describe in issue, I would love to use web hook actions facility to send REST requests to elasticsearch. I clearly prefer to use web hook action than to write manually JavaScript to do that.

**Comment by lmangani** [comment]
@fbaligand sounds similar to the chained events in the ATLAS example we recently got an example of - have you checked if it could be adapted to your purpose? (one watcher inserting data using a webhook, and a second looking through its results in a dedicated index)

**Comment by fbaligand** [comment]
@lmangani 
Yes I read [ATLAS Tutorial in wiki](https://github.com/sirensolutions/sentinl/wiki/Statistical-Anomaly-Detector).
It's a clearly interesting case, when you want to re-aggregate your data and then chain alerting based on this newly aggregated data.

But the use case I speak about is quite different. Because, I want to chain several admin actions on yesterday index. So in this case, I don't produce new aggregated data to be consumed by another watcher by a search query.

The case I speak about would chain the following webhooks :
- `POST logstash-{now-1d/d}/_close`
- `PUT logstash-{now-1d/d}/_settings {" index.codec": "best_compression"}`
- `POST logstash-{now-1d/d}/_open`
- `POST logstash-{now-1d/d}/_forcemerge?max_num_segments=1`

Obviously, I could schedule a first watcher at 0:01, then the second at 0:02 and so on. But honestly, that is tedious.

**Comment by lmangani** [comment]
Thanks for the example, and I agree this is anyway needed to make the tool more useful - my impression is we need to create "one shot" watchers which get executed once and discarded, and provide an action to generate them. This way any watcher could generate another watcher to be executed immediately without passing by the scheduler and disappear once done (or have a child)

**Comment by fbaligand** [comment]
Would not it be easier to just add a "sequential" option on a watcher so that the watcher actions are executed one after the other ?

**Comment by lmangani** [comment]
This would create the need for the watcher to know what other watchers are available and a structure to support that - plus as well the ability to immediately execute them - I think its best to start from this sttep, and have actions generate new watchers directly or using scripts templated by the user (as @chenryn also suggested)

**Comment by fbaligand** [comment]
But I speak about only one watcher and several actions. So there is no relation to manage between several watchers.
In the idea, when watcher's condition is met, first action is executed, then when it's finished (for a webhook, it means that http response is received), the second action is executed, and so on.
And this could be done using javascript promises for example.

**Comment by lmangani** [comment]
so, you're saying each "action" should also have a condition and a script attached right?

**Comment by fbaligand** [comment]
No, I mean :
- 1 watcher 
- 1 condition
- 4 actions (in the specific case I described : all webhooks)

If the condition is met, all 4 actions are executed one after the other.
The only difference compared to the default watcher behavior is that actions are executed sequentially, and not executed in parallel.

Is my explanation clearer ?

**Comment by jccq** [comment]
Maybe having a sort of advanced webhook where one can put more commands?
but the correct handling of responses and all seems.. mmm .. worth a script
externlly executed maybe? ( @fabiocorneti to advise ?)

*Giovanni Tummarello*   Siren Solutions - CEO
M +353 87 1272938 |
giovanni@siren.solutions | http://siren.solutions/
Unit 13 GTC, Mervue Park, Galway, Ireland
<http://www.twitter.com/jccq?utm_source=WiseStamp&utm_medium=email&utm_term=&utm_content=&utm_campaign=signature>
<http://youtube.com/c/SirenSolutionsChannel?utm_source=WiseStamp&utm_medium=email&utm_term=&utm_content=&utm_campaign=signature>
<https://www.linkedin.com/in/giovannitummarello?utm_source=WiseStamp&utm_medium=email&utm_term=&utm_content=&utm_campaign=signature>

**Comment by fbaligand** [comment]
I think it's a good idea !
Thus, among all currently available watcher actions, only webhook would be interested by a "sequential" execution.

So, considering that, I think it would be nice to have a new watcher action called `webhook_chain` (for instance). It would chain (execute sequentially) several webhooks configured as an array.

Here's an example configuration :
``` json
{
    ...
    "actions": {
      "compress_yesterday_index": {
        "webhook_chain": [
            {
                "method" : "POST", 
                "host" : "localhost", 
                "port" : 9200, 
                "path": "/logstash-{now-1d/d}/_close"
            },
            {
                "method" : "PUT", 
                "host" : "localhost", 
                "port" : 9200, 
                "path": "/logstash-{now-1d/d}/_settings", 
                "body": "{\" index.codec\": \"best_compression\"}"
            },
            {
                "method" : "POST", 
                "host" : "localhost", 
                "port" : 9200, 
                "path": "/logstash-{now-1d/d}/_open"
            },
            {
                "method" : "GET", 
                "host" : "localhost", 
                "port" : 9200, 
                "path": "/_cluster/health/logstash-{now-1d/d}?wait_for_status=green"
            },
            {
                "method" : "POST", 
                "host" : "localhost", 
                "port" : 9200, 
                "path": "/logstash-{now-1d/d}/_forcemerge?max_num_segments=1"
            }
        ]
      }
    }
}
```
